packages/remote-feature-flag-controller/src/utils/user-segmentation-utils.ts

```diff
@@ -1,25 +1,45 @@
 import type { Json } from '@metamask/utils';
+import { validate as uuidValidate, version as uuidVersion } from 'uuid';
 
 import type { FeatureFlagScopeValue } from '../remote-feature-flag-controller-types';
 
-/* eslint-disable no-bitwise */
 /**
  * Generates a deterministic random number between 0 and 1 based on a metaMetricsId.
  * This is useful for A/B testing and feature flag rollouts where we want
  * consistent group assignment for the same user.
  *
- * @param metaMetricsId - The unique identifier used to generate the deterministic random number
+ * Supports two metaMetricsId formats:
+ * - UUIDv4 format (new mobile implementation)
+ * - Hex format with 0x prefix (extension or old mobile implementation)
+ *
+ * For UUIDv4 format, the following normalizations are applied:
+ * - Removes all dashes from the UUID
+ * - Remove version (4) bits and replace with 'f'
+ * - Converts the remaining hex string to a BigInt for calculation
+ *
+ * For hex format:
+ * - Expects a hex string with '0x' prefix (e.g., '0x1234abcd')
+ * - Removes the '0x' prefix before conversion
+ * - Converts the remaining hex string to a BigInt for calculation
+ *
+ * @param metaMetricsId - The unique identifier used to generate the deterministic random number, can be a UUIDv4 or hex string
  * @returns A number between 0 and 1 that is deterministic for the given metaMetricsId
  */
 export function generateDeterministicRandomNumber(
   metaMetricsId: string,
 ): number {
-  const hash = [...metaMetricsId].reduce((acc, char) => {
-    const chr = char.charCodeAt(0);
-    return ((acc << 5) - acc + chr) | 0;
-  }, 0);
-
-  return (hash >>> 0) / 0xffffffff;
+  let cleanId: string;
+  // uuidv4 format
+  if (uuidValidate(metaMetricsId) && uuidVersion(metaMetricsId) === 4) {
+    cleanId = metaMetricsId.replace(/-/gu, '').replace(/^(.{12})4/u, '$1f');
+  } else {
+    // hex format with 0x prefix
+    cleanId = metaMetricsId.slice(2);
```

Comment:
Note that this will return the wrong answer if some users have shorter IDs than others.

As far as I can tell this shouldn't be possible. But maybe we should add a guard here to ensure the ID is the length we expect, just so that we get a Sentry error and can adjust this if there is some weird case where IDs are different lengths.

Reply:
And similarly, we might want to throw a more meaningful error if a non-v4 UUID is passed in, or an empty string, or other invalid inputs. Right now it may produce a confusing error.

Reply:
Good catch ! The validation guard is added a66b49ffd980b2fc6231c66d1b043f8ceaaf94ab